File: testcoode/fighter.py
#!/usr/bin/env python3
"""
Fighter Class - D&D 5e SRD Implementation
Derived from the base Character class with full Fighter features
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple


from Head.campaign_manager import Character


@dataclass
class Fighter(Character):
    """
    Fighter class implementation following D&D 5e SRD
    Inherits from Character and adds Fighter-specific features
    """
    
    # Fighter-specific attributes
    martial_archetype: str = ""  # "Champion", "Battle Master", "Eldritch Knight"
    
    # Core Fighter mechanics
    fighting_styles: List[str] = field(default_factory=list)
    second_wind_used: bool = False
    action_surge_uses: int = 0
    action_surge_max_uses: int = 1
    indomitable_uses: int = 0
    indomitable_max_uses: int = 1
    
    # Extra Attack progression
    extra_attacks: int = 1  # Base is 1 attack, increases at levels 5, 11, 20
    
    # Champion archetype features
    improved_critical: bool = False  # Crit on 19-20
    remarkable_athlete: bool = False  # Level 7 Champion
    additional_fighting_style: bool = False  # Level 10 Champion
    superior_critical: bool = False  # Crit on 18-20
    survivor_active: bool = False  # Level 18 Champion
    
    # Archetype features unlocked at certain levels
    archetype_feature_3: bool = False
    archetype_feature_7: bool = False
    archetype_feature_10: bool = False
    archetype_feature_15: bool = False
    archetype_feature_18: bool = False
    
    def __post_init__(self):
        """Initialize Fighter with proper stats and proficiencies"""
        super().__post_init__()
        
        # Set class type
        self.char_class = "Fighter"
        
        # Set initial stats if not provided
        if not self.stats or all(v == 10 for v in self.stats.values()):
            # Fighters typically have high STR and CON
            self.stats = {
                "strength": 16,
                "dexterity": 14,
                "constitution": 15,
                "intelligence": 10,
                "wisdom": 12,
                "charisma": 8
            }
        
        # Set hit points based on level (1d10 per level)
        if self.level == 1:
            self.max_hp = 10 + self._get_modifier("constitution")
            self.hp = self.max_hp
        
        # Initialize proficiencies if empty
        if not self.skill_proficiencies:
            # Fighters choose two from: Acrobatics, Animal Handling, Athletics, 
            # History, Insight, Intimidation, Perception, Survival
            self.skill_proficiencies = ["Athletics", "Intimidation"]
        
        if not self.languages_known:
            self.languages_known = ["Common"]
        
        # Apply level-based features
        self.apply_level_features()
    
    def _get_modifier(self, ability: str) -> int:
        """Calculate ability modifier"""
        return (self.stats.get(ability, 10) - 10) // 2
    
    def apply_level_features(self):
        """Apply features based on current level according to SRD"""
        # Action Surge progression
        if self.level >= 2:
            self.action_surge_max_uses = 1
        if self.level >= 17:
            self.action_surge_max_uses = 2
        
        # Indomitable progression
        if self.level >= 9:
            self.indomitable_max_uses = 1
        if self.level >= 13:
            self.indomitable_max_uses = 2
        if self.level >= 17:
            self.indomitable_max_uses = 3
        
        # Extra Attack progression
        if self.level < 5:
            self.extra_attacks = 1
        elif self.level < 11:
            self.extra_attacks = 2
        elif self.level < 20:
            self.extra_attacks = 3
        else:
            self.extra_attacks = 4
        
        # Class features by level
        if self.level >= 1:
            # Starting fighting style
            if not self.fighting_styles:
                self.fighting_styles = ["Defense"]
        
        if self.level >= 3:
            self.archetype_feature_3 = True
        
        if self.level >= 7:
            self.archetype_feature_7 = True
        
        if self.level >= 10:
            self.archetype_feature_10 = True
        
        if self.level >= 15:
            self.archetype_feature_15 = True
        
        if self.level >= 18:
            self.archetype_feature_18 = True
        
        # Champion archetype features
        if self.martial_archetype == "Champion":
            if self.level >= 3:
                self.improved_critical = True
            if self.level >= 7:
                self.remarkable_athlete = True
            if self.level >= 10:
                self.additional_fighting_style = True
            if self.level >= 15:
                self.superior_critical = True
            if self.level >= 18:
                self.survivor_active = True
    
    def use_second_wind(self) -> Tuple[bool, int]:
        """
        Use Second Wind to regain hit points.
        Returns (success, hp_healed)
        """
        if self.second_wind_used:
            return False, 0
        
        hp_healed = max(1, self.roll_dice(1, 10)) + self.level
        self.hp = min(self.hp + hp_healed, self.max_hp)
        self.second_wind_used = True
        return True, hp_healed
    
    def use_action_surge(self) -> bool:
        """
        Use Action Surge to take an additional action.
        Returns True if successful, False otherwise.
        """
        if self.action_surge_uses >= self.action_surge_max_uses:
            return False
        
        self.action_surge_uses += 1
        return True
    
    def use_indomitable(self) -> bool:
        """
        Use Indomitable to reroll a failed saving throw.
        Returns True if successful, False otherwise.
        """
        if self.indomitable_uses >= self.indomitable_max_uses:
            return False
        
        self.indomitable_uses += 1
        return True
    
    def short_rest(self):
        """Reset short rest resources"""
        self.second_wind_used = False
        self.action_surge_uses = 0
    
    def long_rest(self):
        """Reset all resources on long rest"""
        self.second_wind_used = False
        self.action_surge_uses = 0
        self.indomitable_uses = 0
        self.hp = self.max_hp
    
    def add_fighting_style(self, style: str) -> bool:
        """
        Add a fighting style to the fighter.
        Returns True if successful, False if style already exists.
        """
        if style in self.fighting_styles:
            return False
        
        self.fighting_styles.append(style)
        return True
    
    def get_fighting_style_benefits(self, style: str) -> Dict[str, any]:
        """Get the benefits of a specific fighting style"""
        benefits = {
            "Archery": {
                "description": "+2 bonus to attack rolls with ranged weapons",
                "ranged_attack_bonus": 2
            },
            "Defense": {
                "description": "+1 bonus to AC while wearing armor",
                "ac_bonus": 1
            },
            "Dueling": {
                "description": "+2 bonus to damage rolls with one-handed melee weapons when wielding only one weapon",
                "damage_bonus": 2
            },
            "Great Weapon Fighting": {
                "description": "Reroll 1s and 2s on damage dice for two-handed weapons",
                "reroll_damage": [1, 2]
            },
            "Protection": {
                "description": "Use reaction to impose disadvantage on attack against ally within 5ft when wielding shield",
                "impose_disadvantage": True
            },
            "Two-Weapon Fighting": {
                "description": "Add ability modifier to damage of off-hand attack",
                "add_ability_modifier": True
            }
        }
        
        return benefits.get(style, {})
    
    def calculate_ac_with_fighting_styles(self, base_ac: int) -> int:
        """Calculate AC including fighting style bonuses"""
        ac = base_ac
        if "Defense" in self.fighting_styles:
            ac += 1
        return ac
    
    def get_critical_range(self) -> Tuple[int, int]:
        """Get the critical hit range based on Champion features"""
        if self.superior_critical and self.martial_archetype == "Champion":
            return 18, 20
        elif self.improved_critical and self.martial_archetype == "Champion":
            return 19, 20
        else:
            return 20, 20
    
    def get_remarkable_athlete_bonus(self) -> int:
        """Calculate the Remarkable Athlete bonus (half proficiency, round up)"""
        proficiency_bonus = self.get_proficiency_bonus()
        return (proficiency_bonus + 1) // 2  # Round up
    
    def check_survivor_healing(self) -> int:
        """
        Check if Survivor feature activates at start of turn.
        Returns HP regained (0 if not applicable)
        """
        if not self.survivor_active:
            return 0
        
        if self.hp <= self.max_hp // 2 and self.hp > 0:
            healing = 5 + self._get_modifier("constitution")
            self.hp = min(self.hp + healing, self.max_hp)
            return healing
        
        return 0
    
    def get_proficiency_bonus(self) -> int:
        """Calculate proficiency bonus based on level"""
        if self.level < 5:
            return 2
        elif self.level < 9:
            return 3
        elif self.level < 13:
            return 4
        elif self.level < 17:
            return 5
        else:
            return 6
    
    def level_up(self):
        """Level up the fighter"""
        self.level += 1
        
        # Increase HP (1d10 + CON mod per level, or 6 + CON mod average)
        hp_gain = 6 + self._get_modifier("constitution")
        self.max_hp += hp_gain
        self.hp += hp_gain
        
        # Reapply level-based features
        self.apply_level_features()
    
    def get_character_sheet(self) -> Dict:
        """Generate a complete character sheet"""
        # Calculate AC with fighting styles
        current_ac = self.calculate_ac_with_fighting_styles(self.ac)
        
        return {
            "name": self.name,
            "race": self.race,
            "class": f"{self.char_class} {self.level}",
            "background": self.background,
            "alignment": self.alignment,
            "martial_archetype": self.martial_archetype if self.martial_archetype else "Not chosen (Level 3+)",
            
            "hit_points": f"{self.hp}/{self.max_hp}",
            "armor_class": current_ac,
            "speed": self.speed,
            
            "ability_scores": self.stats,
            "proficiency_bonus": self.get_proficiency_bonus(),
            
            "saving_throws": {
                "strength": self._get_modifier("strength") + self.get_proficiency_bonus(),
                "constitution": self._get_modifier("constitution") + self.get_proficiency_bonus()
            },
            
            "skills": self.skill_proficiencies,
            "languages": self.languages_known,
            
            "combat_features": {
                "fighting_styles": self.fighting_styles,
                "second_wind_available": not self.second_wind_used,
                "action_surge": f"{self.action_surge_uses}/{self.action_surge_max_uses}",
                "indomitable": f"{self.indomitable_uses}/{self.indomitable_max_uses}",
                "extra_attacks": self.extra_attacks
            },
            
            "features": {
                "Fighting Style": ", ".join(self.fighting_styles) if self.fighting_styles else "None chosen",
                "Second Wind": True,
                "Action Surge": self.action_surge_max_uses > 0,
                "Martial Archetype": self.martial_archetype if self.martial_archetype else "Available at Level 3",
                "Extra Attack": f"{self.extra_attacks} attacks" if self.extra_attacks > 1 else False,
                "Indomitable": self.indomitable_max_uses > 0,
                
                # Champion features
                "Improved Critical": self.improved_critical,
                "Remarkable Athlete": self.remarkable_athlete,
                "Additional Fighting Style": self.additional_fighting_style,
                "Superior Critical": self.superior_critical,
                "Survivor": self.survivor_active
            },
            
            "inventory": self.inventory,
            "currency": self.currency,
            "total_wealth_gp": self.get_total_gold_value(),
            
            "personality": {
                "traits": self.personality_traits,
                "ideal": self.ideal,
                "bond": self.bond,
                "flaw": self.flaw
            },
            
            "notes": self.notes
        }
    
    def roll_dice(self, num_dice: int, dice_type: int) -> int:
        """
        Simulate rolling dice. In a real implementation, this would use random.
        For testing purposes, returns average roll.
        """
        return (num_dice * (dice_type + 1)) // 2
    
    def __str__(self) -> str:
        """String representation"""
        archetype_str = f" ({self.martial_archetype})" if self.martial_archetype else ""
        action_surge_str = f" | Action Surge: {self.action_surge_uses}/{self.action_surge_max_uses}"
        second_wind_str = " | Second Wind: Available" if not self.second_wind_used else " | Second Wind: Used"
        return f"{self.name} - Level {self.level} {self.race} {self.char_class}{archetype_str} | HP: {self.hp}/{self.max_hp} | AC: {self.ac}{action_surge_str}{second_wind_str}"


# Example usage and testing
if __name__ == "__main__":
    # Create a new Fighter
    fighter = Fighter(
        name="Sir Reginald Steelheart",
        race="Human",
        char_class="Fighter",
        background="Soldier",
        level=1,
        stats={
            "strength": 16,
            "dexterity": 14,
            "constitution": 15,
            "intelligence": 10,
            "wisdom": 12,
            "charisma": 8
        },
        alignment="Lawful Good",
        personality_traits=["I'm always polite and respectful", "I face problems head-on"],
        ideal="Greater Good. Our lot is to lay down our lives in defense of others",
        bond="I fight for those who cannot fight for themselves",
        flaw="I'd rather eat my armor than admit I'm wrong"
    )
    
    # Set starting equipment
    fighter.inventory = ["Longsword", "Shield", "Chain Mail", "Light Crossbow", "20 Bolts", "Dungeoneer's Pack"]
    
    print("=" * 60)
    print("FIGHTER CHARACTER SHEET")
    print("=" * 60)
    print(fighter)
    print()
    
    # Display character sheet
    sheet = fighter.get_character_sheet()
    print(f"Name: {sheet['name']}")
    print(f"Class: {sheet['class']}")
    print(f"Race: {sheet['race']}")
    print(f"Background: {sheet['background']}")
    print(f"Alignment: {sheet['alignment']}")
    print()
    
    print(f"HP: {sheet['hit_points']} | AC: {sheet['armor_class']} | Speed: {sheet['speed']} ft")
    print()
    
    print("Ability Scores:")
    for ability, score in sheet['ability_scores'].items():
        modifier = (score - 10) // 2
        sign = '+' if modifier >= 0 else ''
        print(f"  {ability.capitalize()}: {score} ({sign}{modifier})")
    print()
    
    print(f"Proficiency Bonus: +{sheet['proficiency_bonus']}")
    print()
    
    print("Combat Features:")
    print(f"  Fighting Styles: {', '.join(sheet['combat_features']['fighting_styles'])}")
    print(f"  Second Wind: {'Available' if sheet['combat_features']['second_wind_available'] else 'Used'}")
    print(f"  Action Surge: {sheet['combat_features']['action_surge']}")
    print(f"  Indomitable: {sheet['combat_features']['indomitable']}")
    print(f"  Extra Attacks: {sheet['combat_features']['extra_attacks']}")
    print()
    
    print("Class Features:")
    for feature, value in sheet['features'].items():
        if value and value is not False:
            print(f"  • {feature}: {value if isinstance(value, str) else '✓'}")
    print()
    
    print("Equipment:")
    for item in fighter.inventory:
        print(f"  • {item}")
    print()
    
    # Test Second Wind
    print("=" * 60)
    print("TESTING SECOND WIND")
    print("=" * 60)
    
    # Take some damage first
    fighter.hp -= 8
    print(f"HP after taking damage: {fighter.hp}/{fighter.max_hp}")
    
    print("\nUsing Second Wind...")
    success, hp_healed = fighter.use_second_wind()
    if success:
        print(f"✓ Second Wind healed {hp_healed} HP!")
        print(f"Current HP: {fighter.hp}/{fighter.max_hp}")
    
    # Test Action Surge
    print("\n" + "=" * 60)
    print("TESTING ACTION SURGE")
    print("=" * 60)
    
    print("\nUsing Action Surge...")
    if fighter.use_action_surge():
        print("✓ Action Surge activated! You can take an additional action this turn.")
        print(f"Action Surge uses remaining: {fighter.action_surge_max_uses - fighter.action_surge_uses}")
    
    # Test leveling up
    print("\n" + "=" * 60)
    print("TESTING LEVEL UP")
    print("=" * 60)
    
    print(f"\nCurrent level: {fighter.level}")
    print(f"Current HP: {fighter.hp}/{fighter.max_hp}")
    print(f"Current extra attacks: {fighter.extra_attacks}")
    
    fighter.level_up()
    fighter.level_up()  # Level up to 3 to choose archetype
    fighter.martial_archetype = "Champion"
    fighter.apply_level_features()
    
    print(f"\nAfter level up to Level {fighter.level}:")
    print(f"New HP: {fighter.hp}/{fighter.max_hp}")
    print(f"Martial Archetype: {fighter.martial_archetype}")
    print(f"Features unlocked: Improved Critical (crit on 19-20)")
    print(f"Critical range: {fighter.get_critical_range()}")
    
    # Level up further to test more features
    for _ in range(7):  # Level up to 10
        fighter.level_up()
    fighter.apply_level_features()
    
    print(f"\nAt Level {fighter.level}:")
    print(f"Extra Attacks: {fighter.extra_attacks}")
    print(f"Action Surge uses: {fighter.action_surge_max_uses}")
    print(f"Features unlocked: Remarkable Athlete, Additional Fighting Style")
    
    # Test Remarkable Athlete
    if fighter.remarkable_athlete:
        bonus = fighter.get_remarkable_athlete_bonus()
        print(f"Remarkable Athlete bonus: +{bonus}")
    
    print("\n" + "=" * 60)
    print(fighter)
    print("=" * 60)

File: testcoode/fighter.js
/**
 * Fighter Class Features UI
 * Handles Fighter-specific displays: Fighting Styles, Second Wind, Action Surge, Feature Tree
 */

class FighterFeatureManager extends ClassFeatureManager {
    constructor() {
        super('Fighter');
        this.secondWindUsed = false;
        this.actionSurgeUses = 0;
        this.actionSurgeMaxUses = 1;
        this.indomitableUses = 0;
        this.indomitableMaxUses = 0;
        this.extraAttacks = 1;
        this.fightingStyles = [];
        this.stats = { strength: 10, dexterity: 10, constitution: 10 };
        this.subclass = '';
    }

    initialize(level, stats, subclass = '', fightingStyles = []) {
        this.level = level;
        this.stats = stats || this.stats;
        this.subclass = subclass;
        this.fightingStyles = fightingStyles || ['Defense']; // Default style
        this.updateProgression();
        this.render();
    }

    updateProgression() {
        // Action Surge progression
        if (this.level >= 17) {
            this.actionSurgeMaxUses = 2;
        } else if (this.level >= 2) {
            this.actionSurgeMaxUses = 1;
        } else {
            this.actionSurgeMaxUses = 0;
        }

        // Indomitable progression
        if (this.level >= 17) {
            this.indomitableMaxUses = 3;
        } else if (this.level >= 13) {
            this.indomitableMaxUses = 2;
        } else if (this.level >= 9) {
            this.indomitableMaxUses = 1;
        } else {
            this.indomitableMaxUses = 0;
        }

        // Extra Attack progression
        if (this.level >= 20) {
            this.extraAttacks = 4;
        } else if (this.level >= 11) {
            this.extraAttacks = 3;
        } else if (this.level >= 5) {
            this.extraAttacks = 2;
        } else {
            this.extraAttacks = 1;
        }
    }

    calculateACWithFightingStyle(baseAC) {
        let ac = baseAC;
        if (this.fightingStyles.includes('Defense')) {
            ac += 1;
        }
        return ac;
    }

    getCriticalRange() {
        if (this.subclass === 'Champion') {
            if (this.level >= 15) return '18-20';
            if (this.level >= 3) return '19-20';
        }
        return '20';
    }

    getRemarkableAthleteBonus() {
        const proficiencyBonus = this.getProficiencyBonus();
        return Math.ceil(proficiencyBonus / 2);
    }

    getProficiencyBonus() {
        if (this.level < 5) return 2;
        if (this.level < 9) return 3;
        if (this.level < 13) return 4;
        if (this.level < 17) return 5;
        return 6;
    }

    render() {
        const container = document.getElementById('fighter-features');
        if (!container) return;

        container.innerHTML = `
            <div class="class-specific-section active">
                <div class="class-section-header">
                     Fighter Features
                </div>
                
                ${this.renderResourceTrackers()}
                ${this.renderFightingStyles()}
                ${this.renderFeatureTree()}
            </div>
        `;

        this.attachEventListeners();
    }

    renderResourceTrackers() {
        const secondWindPercentage = this.secondWindUsed ? 0 : 100;
        const actionSurgePercentage = ((this.actionSurgeMaxUses - this.actionSurgeUses) / this.actionSurgeMaxUses) * 100;
        const indomitablePercentage = ((this.indomitableMaxUses - this.indomitableUses) / this.indomitableMaxUses) * 100;

        return `
            <div class="resource-trackers">
                <!-- Second Wind -->
                <div class="resource-tracker">
                    <div class="resource-name">
                        💨 Second Wind
                    </div>
                    <div class="resource-bar">
                        <div class="resource-fill" style="width: ${secondWindPercentage}%; background: linear-gradient(90deg, #4ecdc4 0%, #44a08d 100%);">
                            ${this.secondWindUsed ? 'Used' : 'Available'}
                        </div>
                    </div>
                    <div class="resource-buttons">
                        <button id="second-wind-btn" ${this.secondWindUsed ? 'disabled' : ''}>
                            Use Second Wind
                        </button>
                        <button id="short-rest-btn">
                            Short Rest
                        </button>
                    </div>
                </div>

                <!-- Action Surge -->
                <div class="resource-tracker">
                    <div class="resource-name">
                        ⚡ Action Surge
                    </div>
                    <div class="resource-bar">
                        <div class="resource-fill" style="width: ${actionSurgePercentage}%; background: linear-gradient(90deg, #ffd93d 0%, #ffb347 100%);">
                            ${this.actionSurgeUses}/${this.actionSurgeMaxUses}
                        </div>
                    </div>
                    <div class="resource-buttons">
                        <button id="action-surge-btn" ${this.actionSurgeUses >= this.actionSurgeMaxUses ? 'disabled' : ''}>
                            Use Action Surge
                        </button>
                        <button id="short-rest-btn-2">
                            Short Rest
                        </button>
                    </div>
                </div>

                <!-- Indomitable -->
                ${this.indomitableMaxUses > 0 ? `
                <div class="resource-tracker">
                    <div class="resource-name">
                        🛡️ Indomitable
                    </div>
                    <div class="resource-bar">
                        <div class="resource-fill" style="width: ${indomitablePercentage}%; background: linear-gradient(90deg, #6a89cc 0%, #4a69bd 100%);">
                            ${this.indomitableUses}/${this.indomitableMaxUses}
                        </div>
                    </div>
                    <div class="resource-buttons">
                        <button id="indomitable-btn" ${this.indomitableUses >= this.indomitableMaxUses ? 'disabled' : ''}>
                            Use Indomitable
                        </button>
                        <button id="long-rest-btn">
                            Long Rest
                        </button>
                    </div>
                </div>
                ` : ''}
            </div>
        `;
    }

    renderFightingStyles() {
        const styleBenefits = {
            'Archery': '+2 to attack rolls with ranged weapons',
            'Defense': '+1 AC while wearing armor',
            'Dueling': '+2 damage with one-handed melee weapons (no other weapons)',
            'Great Weapon Fighting': 'Reroll 1s and 2s on damage dice for two-handed weapons',
            'Protection': 'Use reaction to impose disadvantage on attack against ally within 5ft (shield required)',
            'Two-Weapon Fighting': 'Add ability modifier to damage of off-hand attack'
        };

        let stylesHtml = this.fightingStyles.map(style => `
            <div class="fighting-style-card">
                <div class="fighting-style-name">${style}</div>
                <div class="fighting-style-desc">${styleBenefits[style] || 'No description available'}</div>
            </div>
        `).join('');

        // Show additional fighting style option for Champion at level 10
        const canChooseAdditionalStyle = this.subclass === 'Champion' && 
                                       this.level >= 10 && 
                                       this.fightingStyles.length < 2;

        if (canChooseAdditionalStyle) {
            stylesHtml += `
                <div class="fighting-style-selector">
                    <div class="fighting-style-name">➕ Additional Fighting Style</div>
                    <select id="additional-style-select">
                        <option value="">Choose a second fighting style...</option>
                        ${Object.keys(styleBenefits).map(style => 
                            !this.fightingStyles.includes(style) ? 
                            `<option value="${style}">${style}</option>` : ''
                        ).join('')}
                    </select>
                    <button id="add-style-btn">Add Style</button>
                </div>
            `;
        }

        // Show critical range for Champion
        let championFeatures = '';
        if (this.subclass === 'Champion') {
            const critRange = this.getCriticalRange();
            championFeatures = `
                <div class="champion-features">
                    <div class="calculation-display">
                        <div class="ability-card-name">🎯 Improved Critical</div>
                        <div class="calculation-result">
                            Critical Hit: ${critRange}
                        </div>
                    </div>
                    ${this.level >= 7 ? `
                    <div class="calculation-display">
                        <div class="ability-card-name">🏃 Remarkable Athlete</div>
                        <div class="calculation-formula">
                            +${this.getRemarkableAthleteBonus()} to STR/DEX/CON checks without proficiency
                        </div>
                    </div>
                    ` : ''}
                    ${this.level >= 18 ? `
                    <div class="calculation-display">
                        <div class="ability-card-name">❤️ Survivor</div>
                        <div class="calculation-formula">
                            Regain 5 + CON mod HP at start of turn if below half HP
                        </div>
                    </div>
                    ` : ''}
                </div>
            `;
        }

        return `
            <div class="fighting-styles-section">
                <div class="section-header">Fighting Styles</div>
                <div class="fighting-styles-grid">
                    ${stylesHtml}
                </div>
                ${championFeatures}
            </div>
        `;
    }

    renderFeatureTree() {
        const features = [
            { level: 1, name: 'Fighting Style', desc: this.fightingStyles.join(', ') || 'Choose a style' },
            { level: 1, name: 'Second Wind', desc: 'Bonus action to heal 1d10 + level, once per short rest' },
            { level: 2, name: 'Action Surge', desc: `Take one additional action, ${this.actionSurgeMaxUses} use(s) per short rest` },
            { level: 3, name: 'Martial Archetype', desc: this.subclass || 'Choose your archetype' },
            { level: 4, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 5, name: 'Extra Attack', desc: `Attack ${this.extraAttacks} times when you take the Attack action` },
            { level: 6, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 7, name: 'Archetype Feature', desc: this.getArchetypeFeatureDesc(7) },
            { level: 8, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 9, name: 'Indomitable', desc: `Reroll a failed saving throw, ${this.indomitableMaxUses} use(s) per long rest` },
            { level: 10, name: 'Archetype Feature', desc: this.getArchetypeFeatureDesc(10) },
            { level: 11, name: 'Extra Attack (2)', desc: 'Attack 3 times when you take the Attack action' },
            { level: 12, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 13, name: 'Indomitable (2 uses)', desc: 'Reroll failed saves twice per long rest' },
            { level: 14, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 15, name: 'Archetype Feature', desc: this.getArchetypeFeatureDesc(15) },
            { level: 16, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 17, name: 'Action Surge (2 uses)', desc: 'Use Action Surge twice per short rest' },
            { level: 17, name: 'Indomitable (3 uses)', desc: 'Reroll failed saves three times per long rest' },
            { level: 18, name: 'Archetype Feature', desc: this.getArchetypeFeatureDesc(18) },
            { level: 19, name: 'Ability Score Improvement', desc: '+2 to one ability or +1 to two abilities' },
            { level: 20, name: 'Extra Attack (3)', desc: 'Attack 4 times when you take the Attack action' }
        ];

        let html = '<div class="feature-tree">';
        
        features.forEach(feature => {
            const unlocked = this.isFeatureUnlocked(feature.level);
            html += `
                <div class="feature-node ${unlocked ? 'unlocked' : 'locked'}">
                    <div class="feature-level">Lv ${feature.level}</div>
                    <div style="flex: 1;">
                        <div class="feature-name">${feature.name}</div>
                        <div class="feature-desc">${feature.desc}</div>
                    </div>
                </div>
            `;
        });

        html += '</div>';
        return html;
    }

    getArchetypeFeatureDesc(level) {
        if (this.subclass === 'Champion') {
            switch(level) {
                case 3: return 'Improved Critical: Critical hit on 19-20';
                case 7: return 'Remarkable Athlete: Half proficiency to STR/DEX/CON checks';
                case 10: return 'Additional Fighting Style: Choose a second style';
                case 15: return 'Superior Critical: Critical hit on 18-20';
                case 18: return 'Survivor: Regain HP when below half HP';
                default: return 'Archetype feature';
            }
        } else if (this.subclass === 'Battle Master') {
            switch(level) {
                case 3: return 'Combat Superiority: Learn maneuvers and superiority dice';
                case 7: return 'Know Your Enemy: Insight into creature abilities';
                case 10: return 'Improved Combat Superiority: d10 superiority dice';
                case 15: return 'Relentless: Regain superiority dice when initiative rolled';
                case 18: return 'd12 superiority dice';
                default: return 'Archetype feature';
            }
        } else if (this.subclass === 'Eldritch Knight') {
            switch(level) {
                case 3: return 'Spellcasting: Learn wizard spells';
                case 7: return 'War Magic: Cast cantrip and make weapon attack';
                case 10: return 'Eldritch Strike: Disadvantage on saves against your spells';
                case 15: return 'Arcane Charge: Teleport when using Action Surge';
                case 18: return 'Improved War Magic: Attack after casting spell';
                default: return 'Archetype feature';
            }
        }
        return 'Choose archetype at level 3';
    }

    attachEventListeners() {
        // Second Wind
        const secondWindBtn = document.getElementById('second-wind-btn');
        if (secondWindBtn) {
            secondWindBtn.addEventListener('click', () => this.useSecondWind());
        }

        // Action Surge
        const actionSurgeBtn = document.getElementById('action-surge-btn');
        if (actionSurgeBtn) {
            actionSurgeBtn.addEventListener('click', () => this.useActionSurge());
        }

        // Indomitable
        const indomitableBtn = document.getElementById('indomitable-btn');
        if (indomitableBtn) {
            indomitableBtn.addEventListener('click', () => this.useIndomitable());
        }

        // Short Rest buttons
        const shortRestBtns = document.querySelectorAll('#short-rest-btn, #short-rest-btn-2');
        shortRestBtns.forEach(btn => {
            if (btn) btn.addEventListener('click', () => this.shortRest());
        });

        // Long Rest
        const longRestBtn = document.getElementById('long-rest-btn');
        if (longRestBtn) {
            longRestBtn.addEventListener('click', () => this.longRest());
        }

        // Additional Fighting Style (Champion)
        const addStyleBtn = document.getElementById('add-style-btn');
        if (addStyleBtn) {
            addStyleBtn.addEventListener('click', () => this.addAdditionalStyle());
        }
    }

    useSecondWind() {
        if (this.secondWindUsed) return;

        const healing = Math.floor(Math.random() * 10) + 1 + this.level; // 1d10 + level
        this.secondWindUsed = true;
        this.render();
        
        this.showNotification(`💨 Second Wind healed ${healing} HP!`);
        
        // In a real implementation, you'd update the character's HP here
        // this.character.hp = Math.min(this.character.hp + healing, this.character.maxHp);
    }

    useActionSurge() {
        if (this.actionSurgeUses >= this.actionSurgeMaxUses) return;

        this.actionSurgeUses++;
        this.render();
        
        this.showNotification('⚡ Action Surge activated! Take an additional action this turn.');
    }

    useIndomitable() {
        if (this.indomitableUses >= this.indomitableMaxUses) return;

        this.indomitableUses++;
        this.render();
        
        this.showNotification('🛡️ Indomitable activated! Reroll a failed saving throw.');
    }

    shortRest() {
        this.secondWindUsed = false;
        this.actionSurgeUses = 0;
        this.render();
        
        this.showNotification('✨ Short rest completed. Second Wind and Action Surge restored!');
    }

    longRest() {
        this.secondWindUsed = false;
        this.actionSurgeUses = 0;
        this.indomitableUses = 0;
        this.render();
        
        this.showNotification('🌙 Long rest completed. All features restored!');
    }

    addAdditionalStyle() {
        const select = document.getElementById('additional-style-select');
        if (!select || !select.value) return;

        const newStyle = select.value;
        if (this.fightingStyles.includes(newStyle)) {
            this.showNotification('You already have that fighting style!');
            return;
        }

        this.fightingStyles.push(newStyle);
        this.render();
        
        this.showNotification(`🎯 Added ${newStyle} fighting style!`);
    }

    showNotification(message) {
        // Simple notification - could be enhanced with a toast system
        const notification = document.createElement('div');
        notification.className = 'feature-unlock-notice';
        notification.textContent = message;
        notification.style.position = 'fixed';
        notification.style.top = '20px';
        notification.style.right = '20px';
        notification.style.zIndex = '1000';
        notification.style.animation = 'slideIn 0.3s ease-out';
        
        document.body.appendChild(notification);
        
        setTimeout(() => {
            notification.style.animation = 'slideOut 0.3s ease-out';
            setTimeout(() => notification.remove(), 300);
        }, 3000);
    }

    updateStats(stats) {
        this.stats = stats;
        this.render();
    }

    setLevel(level) {
        this.level = parseInt(level) || 1;
        this.updateProgression();
        this.render();
    }

    setSubclass(subclass) {
        this.subclass = subclass;
        this.render();
    }

    setFightingStyles(styles) {
        this.fightingStyles = styles;
        this.render();
    }
}

// Export for global use
window.FighterFeatureManager = FighterFeatureManager;

// Auto-initialize if on character creation page
document.addEventListener('DOMContentLoaded', () => {
    const classSelect = document.getElementById('char-class');
    if (classSelect && classSelect.value === 'Fighter') {
        window.fighterManager = new FighterFeatureManager();
    }
});